lib/Sema/CSSimplify.cpp

```diff
@@ -6948,16 +6948,23 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
       Type optTy = TypeChecker::getOptionalType(SourceLoc(), innerTV);
       assert(!optTy->hasError());
       SmallVector<Constraint *, 2> optionalities;
-      auto nonoptionalResult = Constraint::createFixed(
-          *this, ConstraintKind::Bind,
-          UnwrapOptionalBase::create(*this, member, locator), memberTy, innerTV,
-          locator);
-      auto optionalResult = Constraint::createFixed(
-          *this, ConstraintKind::Bind,
-          UnwrapOptionalBase::createWithOptionalResult(*this, member, locator),
-          optTy, memberTy, locator);
-      optionalities.push_back(nonoptionalResult);
-      optionalities.push_back(optionalResult);
+      if (flags.contains(TMF_ApplyingFix)) {
+        optionalities.push_back(Constraint::create(*this, ConstraintKind::Bind,
+                                                   memberTy, innerTV, locator));
```

Comment:
What case is this handling? I'm not sure we should be silently unwrapping the base without recording a fix for it.

Reply:
But as for my understanding, we shouldn't attempt fixes when type matching is `TMF_ApplyingFix`  ... this handles the case 

```swift
class T {
    func m1() {
      let l = self.m2!.prop1
    }

    func m2() -> U! {
      return U()
    }
}
```
Where if `m2` is a function type with no argument we attempt to simplify member with the result type `U?` as  base to verify if it has a `prop1` member in order to suggest if this could have a "make it a call" fix `self.m2()!.prop1` in this case. In this case, we only want to match the types and not record fixes(that is why I think the `TMF_ApplyingFix` is passed) which will end up in the wrong diagnostic, so that is why I think this is fine because it will still record fixes in normal no `TMF_ApplyingFix` case.

Reply:
@LucianoPAlmeida Ah I see! 

I don't think using `TMF_ApplyingFix` here is the right fix, it's only really used when additional fixes are unlikely to be useful (and in such cases it returns `SolutionKind::Error` if a fix would be required). IMO in an ideal world we wouldn't need `TMF_ApplyingFix` at all, as we should be able to rely on ranking fixes by their impact.

For the example you give, it seems to me that it might be worth attempting to record an `InsertExplicitCall` fix within `simplifyOptionalObjectConstraint` if we're trying to perform an optional unwrap on a function type of the form `() -> T?`, which would allow us to re-write the code as `self.m2()!.prop1` without telling the user add or remove a `!`.

Reply:
> For the example you give, it seems to me that it might be worth attempting to record an InsertExplicitCall fix within simplifyOptionalObjectConstraint if we're trying to perform an optional unwrap on a function type of the form () -> T?, which would allow us to re-write the code as self.m2()!.prop1 without telling the user add or remove a !.


Right thank you @hamishknight, I'll have to think about this ... TBH I still quite didn't fully understand. Because even doing this we still have to verify if the return of `()-> U?` in this case `U?` has a `prop1` member in order to know if make it a call `()` is applicable, which still would be trying to `simplifyMemberConstraint` with `U?` as a base and would lead to wrong diagnostics when failing on any of the fixed constraints... I think I just have to think about this a bit more, thanks @hamishknight

Reply:
@hamishknight Since the only thing we are doing here is to verify if `U` from result type `U?` has a member, I think we could do is strip the optional out for the check because this shouldn't matter in this situation and we avoid producing the misleading diagnostic. I just updated the last commit, let me know what you think :)

Reply:
> Because even doing this we still have to verify if the return of `()-> U?` in this case `U?` has a `prop1` member in order to know if make it a call `()` is applicable, which still would be trying to `simplifyMemberConstraint` with `U?` as a base and would lead to wrong diagnostics when failing on any of the fixed constraints

You should be able to unwrap the result type of the function when applying the `InsertExplicitCall` fix in `simplifyOptionalObjectConstraint` and match it against the type `second`, which will allow it to propagate to the member constraint which should then succeed with the fix in place.

Reply:
Right, I can take a look on this :)
But this would mean we have to remove the logic from `simplifyMemberConstraint` that currently diagnose this?

Reply:
@LucianoPAlmeida This would be in addition to the logic in `simplifyMemberConstraint`. Trying the fix in `simplifyOptionalObjectConstraint` would handle general cases where a function call needs to be inserted before the unwrap e.g `x()!` or `x()?.y`, whereas the logic in `simplifyMemberConstraint` handles cases where a function call needs to be inserted before the member access such as `x!().y`.

We could try to reduce the duplicated logic by factoring it out into an `InsertExplicitCall::attempt` method (which could take a parameter determining whether the function result needs unwrapping).

Reply:
Ahh I see! Thanks @hamishknight, I think I got it now. Will send a patch ASAP

Reply:
> You should be able to unwrap the result type of the function when applying the `InsertExplicitCall` fix in `simplifyOptionalObjectConstraint` and match the unwrapped type against the type `second`, which will allow it to propagate to the member constraint which should then succeed with the fix in place.

What do you mean by match? I'm just trying to wrap my head around how the member constraint would be propagated here.

Reply:
@LucianoPAlmeida You'll want to call `matchTypes` and check if the result is a success before recording the fix, similar to how it is done in `repairFailures` (though you'll want a `ConstraintKind::Bind` instead of `ConstraintKind::Conversion`:

https://github.com/apple/swift/blob/c825d7987eba360c77e8343e106addaadad4491e/lib/Sema/CSSimplify.cpp#L3141-L3146

The propagation of the base type should then happen automatically.

Reply:
Hum, that is what I thought, in fact, that is what I'm doing, so I think I'm not totally lost as I thought on this one. Thank you @hamishknight :)

Reply:
@hamishknight Adjusted 👍

Reply:
Thanks! I have a bit more feedback, but otherwise this is looking great :)